python/tchannel/statsd.py

```diff
@@ -0,0 +1,143 @@
+
+    def after_receive_system_error(self, request, error):
+        statsd_name = "tchannel.outbound.calls.system-errors"
+        (service, target_service, target_endpoint) = (
+            extract_meta_info_from_request(request))
+        key = '.'.join([statsd_name,
+                        clean(service, 'service'),
+                        clean(target_service, 'target-service'),
+                        clean(target_endpoint, 'target-endpoint'),
+                        clean(ErrorMessage.ERROR_CODES.
+                              get(error.code, None), 'type')
+                        ])
+
+        self._statsd.count(key, 1)
+
+    def after_receive_system_error_per_attempt(self, request, error):
+        statsd_name = "tchannel.outbound.calls.per-attempt.system-errors"
+        (service, target_service, target_endpoint) = (
+            extract_meta_info_from_request(request))
+        key = '.'.join([statsd_name,
+                        clean(service, 'service'),
+                        clean(target_service, 'target-service'),
+                        clean(target_endpoint, 'target-endpoint'),
+                        clean(ErrorMessage.ERROR_CODES.
+                              get(error.code, None), 'type')
+                        ])
+
+        self._statsd.count(key, 1)
+
+    def on_operational_error_per_attempt(self, request, error):
+        statsd_name = "tchannel.outbound.calls.per-attempt.operational-errors"
+        (service, target_service, target_endpoint) = (
+            extract_meta_info_from_request(request))
+        key = '.'.join([statsd_name,
+                        clean(service, 'service'),
+                        clean(target_service, 'target-service'),
+                        clean(target_endpoint, 'target-endpoint'),
+                        clean(ErrorMessage.ERROR_CODES.
+                              get(error.code, None), 'type')
+                        ])
+
+        self._statsd.count(key, 1)
+
+    def on_operational_error(self, request, error):
+        statsd_name = "tchannel.outbound.calls.operational-errors"
+        (service, target_service, target_endpoint) = (
+            extract_meta_info_from_request(request))
+        key = '.'.join([statsd_name,
+                        clean(service, 'service'),
+                        clean(target_service, 'target-service'),
+                        clean(target_endpoint, 'target-endpoint'),
+                        clean(ErrorMessage.ERROR_CODES.
+                              get(error.code, None), 'type')
+                        ])
+
+        self._statsd.count(key, 1)
+
+
+def extract_meta_info_from_request(request):
+    service = request.headers.get('cn', None)
+    target_service = request.service
+    target_endpoint = request.endpoint
+
+    return (service, target_service, target_endpoint)
+
+
+def clean(key, field):
+    if not key:
+        return 'no-' + field
+    else:
+
+        return key.replace(
```

Comment:
There is a performance comparison http://stackoverflow.com/questions/3411771/multiple-character-replace-with-python